base/iterators.jl

```diff
@@ -649,6 +649,51 @@ end
 iterate(it::Drop, state) = iterate(it.xs, state)
 isdone(it::Drop, state) = isdone(it.xs, state)
 
+
+# takewhile
+
+struct TakeWhile{I,P<:Function}
+    xs::I
+    pred::P
+end
+
+takewhile(pred,xs) = TakeWhile(xs,pred)
+
+function iterate(ibl::TakeWhile{I,P},itr...) where {I,P}
+    y = iterate(ibl.xs,itr...)
+    y === nothing && return nothing
+    ibl.pred(y[1]) || return nothing
+    y
+end
+
+IteratorSize(ibl::TakeWhile{I,P}) where {I,P} = SizeUnknown()
+eltype(::Type{TakeWhile{I,P}}) where {I,P} = eltype(I)
+IteratorEltype(::Type{TakeWhile{I,P}}) where {I,P} = IteratorEltype(I)
+
+
+# dropwhile
+
+struct DropWhile{I,P<:Function}
+    xs::I
+    pred::P
+end
+
+dropwhile(pred,itr) = DropWhile(itr,pred)
+function iterate(ibl::DropWhile{I,P},itr=nothing) where {I,P}
+    (fltd,itr) = itr===nothing ? (false,()) : itr
+
+    y = fltd ?
+        iterate(ibl.xs, itr...) :
+        iterate(Iterators.filter(!ibl.pred, ibl.xs),itr...)
+
+    y !== nothing ? (y[1],(true,y[2])) : nothing
```

Comment:
Hmm, I was wrong before --- this doesn't even need to be part of the state, and can be handled simply by doing a while loop at the beginning of iteration.

Reply:
all good, i had a test failure and it is a race that happens when two dropwhile calls were nested. this should help to fix. i have some work ahead for now, i will fix it tomorrow.